### ISSUE TITLE
Question about merge_method

### ISSUE THREAD

**Issue by holmescharles** [spyking-circus/spyking-circus#266]
I have a linear probe with 32 contacts spaced 100 um apart. 16 contacts are dead. Radius = 90 um, so no spatial information should be used for sorts. I have about 3-4 hours of data.

I sorted the data with Syking Circus under using both the distance method and the nd-bhatta method for the `merge_method`. In both cases I used the default parameter. "Distance" yielded ~40 clusters, but nd-bhatta yielded ~250.

For now, I will stick with distance as I believe 50 is a more reasonable (~3-4 clusters per channel), whereas 250 seems erroneous (15-16 clusters per channel). But am I thinking about this wrong? What is a reasonable number of clusters to get out? Is fewer clusters per channel a sign of a bad sort? Furthermore, is there some benefit to using nd-bhatta that I would be missing out on? Should I tweak the parameter to lower the number of clusters out?

**Comment by yger** [comment]
Indeed, this is strange, because from benchmarks the methods do not behave so differently. The high number of cluster persists even after merging, at the end? Because I agree with you that 3-4 clusters per channel is much more reasonable, so in this case it means that nd-bhatta is failing. To understand more in depth what is happening, you can set make_plots = png in the [clustering] section, and have a look to the clustering plot. Maybe you can share some of them here with me? As said, both methods should be rather equivalent, nd-bhatta being usually slightly more sensitive (but in your case, this is clearly too much).

**Comment by SaraMati** [comment]
we had the same issue. The implementation of nd-bhatta method is faulty. we reverted to 'bhatta' and that is fine, i.e. reasonable number of clusters.

**Comment by yger** [comment]
you were all using more or less independent channels ? Maybe nd-bhatta is less efficient for such a configuration, I'll make some more tests. But yes, otherwise you can safely revert to distance or bhatta

**Comment by SaraMati** [comment]
yes, ours is MEA with 200 per electrode distancing in the probe design and I chose 100 for the radius.

**Comment by holmescharles** [comment]
Ditto. By construction my radius was shorter than my inter-channel distance (90 um vs 100 um).

I have not tried bhatta. Is there a link or resource you might be able to point me to so I can understand the difference between nd-bhatta and bhatta?

I can get some data to you. Just need to get data to my advisor first!

Thanks!

**Comment by alihaydaroglu** [comment]
On the same data as @SaraMati, using the nd-bhatta merging method consistently yielded 0 local merges in the clustering step. This was consistent across multiple runs of the package with merging parameters 1, 1.5, 2, and 5. This lead to an unreasonably high number of clusters at the end of the clustering. Changing the method to bhatta significantly improved it, reducing clusters from 1000+ to ~150.

**Comment by alihaydaroglu** [comment]
To look into it a little further, I checked the outputs of the `nd_bhatta_dist` function in `shared/utils.py`. The variables `det_1` and `det_2` (the calculated determinants of the covariance matrices of X1 and X2) are always 0.0 or `nan` each time the function is called. This results in the final distance being `nan`, which I believe is the reason no merges are performed. I'm not sure what about our specific dataset results in the `nan`s. 

For now, I will continue using the `bhatta` distance metric which seems to work fine. Could you explain the difference between the `nd-bhatta` and `bhatta` methods, specifically what the `nd` stands for? 

Thanks for the great software package!

**Comment by yger** [comment]
The Bhattacharyya distance is used to distinguish if two distributions are the same or not. While you have a simple case for 1D distribution (bhatta), it can also be extended to the case of N-dimensional distributions (nd-bhatta) https://en.wikipedia.org/wiki/Bhattacharyya_distance
While this seems to work fine in multiple benchmarks we are running using the SpikeForest toolbox, it looks like you have dataset where this is failing, and we must correct that. I'll have a look and try to make a patch, spotting the failures and fixing this behavior

**Comment by alihaydaroglu** [comment]
Thank you @yger ! Let me know if you need any help reproducing the problem, I'm happy to poke around a little in the source code while running with my data if you need it.

**Comment by AliceFalck** [comment]
Hello! I am working with a similar electrode layout, my electrode distance is also 200 µm and I set the radius to 100. And I can also not use the nd-bhatta. since this issue came up about a year ago and I am using the SC Version 1.0.9, I was wondering if you made any more changes concerning this issue?
